concent_api/core/message_handlers.py

```diff
@@ -751,6 +752,15 @@ def handle_send_force_payment(
             reason = message.concents.ServiceRefused.REASON.DuplicateRequest
         )
 
+    for subtask_results_accepted in client_message.subtask_results_accepted_list:
+        try:
+            validate_task_to_compute(subtask_results_accepted.task_to_compute)
```

Comment:
@pawelkisielewicz All `subtask_results_accepted.task_to_compute` must go through full validation here. They come from the client and the field is specific to this particular use case so we're definitely not validating it in decorators before this point.

It seems to me that `NonPositivePriceTaskToComputeError` is being caught here because all prices being zero is a special case that's not an error. Instead we must tell the client that there's nothing to pay.

Any other validation errors are handled by `@common.decorators.require_golem_message`.

@rwrzesien But I agree with @pawelkisielewicz  that using this validation to detect the case of price being zero is not a good idea. It's too broad. I can see two bugs here:
- The handler returns `NoUnsettledTasksFound` when the price of only some tasks is zero. E.g. if prices are `[100, 200, 0, 300]`. We should proceed with the settlement rather than return in such a case.
- The handler returns `NoUnsettledTasksFound` when there are negative prices. E.g. if prices are `[100, 200, 5, -300]`. It should be a validation error.

Reply:
>  I assume that if flow came to this point task_to_compute was previous validated and everything is OK, but I'm not sure.

No, it wasn't, that is why I have added this part.

> If it was not validated and there is risk that validation may fail because of other reason add handling of this exception

It is already handled in decorators, like for the others message handlers flow. So what I am doing here is 
1. Validating all nested `task_to_compute`.
2a. If `price==0` error is found, handle it in special way (which is done in a way to keep the same API as before those changes)
2b. If any other problem is found, handle it in regular way (through decorator).

> the validation itself should allow price == 0 because a zero price is not a validation error 

We discussed this with @shadeofblue and agreed that `task_to_compute` with `price==0` is wrong from Concent point of view and should be always refused, this validation is now in every use case.

Reply:
Is it wrong from Concent's point of view or wrong in general? For me a validation error means that the message is malformed. It can't even be processed and should be immediately rejected. Message is either valid or not. It should not be valid for one client and invalid for another one.

If it's wrong only for Concent then in my opinion it's business logic and it should be checked separately.

> 2a. If price==0 error is found, handle it in special way (which is done in a way to keep the same API as before those changes)

Where does this special case come from?

Reply:
In previous code there was a case when total price of summed tasks was 0, then `ForcePaymentRejected` with reason `NoUnsettledTasksFound` was returned. I thought it would be good to extend it to catch any task with 0 price.

Reply:
Done as a separate validation, but this part of code will never be reached .

Reply:
Yes, and I what I'm saying does not contradict that. I'm simply saying that the function that validates `TaskToCompute` is not the right place to refuse it because a message with price=0 is not malformed. It's rejected for a different reason.